packages/fdc3-agent-proxy/src/messaging/AbstractMessaging.ts

```diff
@@ -1,165 +1,91 @@
-import { AppIdentifier, ImplementationMetadata, GetAgentParams } from '@kite9/fdc3-standard';
+import { AppIdentifier } from '@kite9/fdc3-standard';
 import { Messaging } from '../Messaging';
 import { RegisterableListener } from '../listeners/RegisterableListener';
-import {
-  AppRequestMessage,
-  WebConnectionProtocol4ValidateAppIdentity,
-  WebConnectionProtocol5ValidateAppIdentitySuccessResponse,
-} from '@kite9/fdc3-schema/generated/api/BrowserTypes';
-type RequestMetadata = AppRequestMessage['meta'];
+import { AgentResponseMessage, AppRequestMessage } from '@kite9/fdc3-schema/generated/api/BrowserTypes';
 
 export abstract class AbstractMessaging implements Messaging {
-  private readonly options: GetAgentParams;
-  private readonly connectionAttemptUuid: string;
-  private readonly timeout: number;
-  private readonly actualUrl: string;
-  private appId: AppIdentifier | null = null;
-  private implementationMetadata: ImplementationMetadata | null = null;
+  // private implementationMetadata: ImplementationMetadata;
+  private appIdentifier: AppIdentifier;
 
   abstract createUUID(): string;
   abstract post(message: object): Promise<void>;
 
   abstract register(l: RegisterableListener): void;
   abstract unregister(id: string): void;
 
-  abstract createMeta(): RequestMetadata;
+  abstract createMeta(): AppRequestMessage['meta'];
 
-  constructor(options: GetAgentParams, connectionAttemptUuid: string, actualUrl: string, timeout: number = 10016) {
-    this.options = options;
-    this.connectionAttemptUuid = connectionAttemptUuid;
-    this.timeout = timeout;
-    this.actualUrl = actualUrl;
-  }
+  abstract getTimeoutMs(): number;
 
-  getSource(): AppIdentifier {
-    if (this.appId === null) {
-      throw new Error('Unconnected. No source available.');
-    }
-    return this.appId;
+  constructor(appIdentifier: AppIdentifier) {
+    this.appIdentifier = appIdentifier;
   }
 
-  waitFor<X>(filter: (m: any) => boolean, timeoutErrorMessage?: string): Promise<X> {
+  waitFor<X extends AgentResponseMessage>(
+    filter: (m: AgentResponseMessage) => boolean,
+    timeoutErrorMessage?: string
+  ): Promise<X> {
     const id = this.createUUID();
     return new Promise<X>((resolve, reject) => {
       let done = false;
-      const listener: RegisterableListener = {
+      let timeout: NodeJS.Timeout | null = null;
+      const l: RegisterableListener = {
         id,
         filter: filter,
         action: m => {
           done = true;
           this.unregister(id);
+          if (timeout) {
+            clearTimeout(timeout);
+          }
           resolve(m);
         },
         register: async () => {
-          this.register(listener);
+          this.register(l);
         },
         unsubscribe: async () => {
           this.unregister(id);
         },
       };
 
-      listener.register();
+      this.register(l);
```

Comment:
Just flagging this line because `this` scoping is a common pain point in javascript. I think `this` refers to the promise at this point, but could refer to the class...

Either way, please double check this line to make absolutely certain you want `this.register(l)` instead of `l.register`

Reply:
We're outside the promise and inside the top-level of the `waitFor` implementation here, so should be correct. Soooo many tests would fail if not.

The `register` and `unsubscribe` inside of the waitFor implementation could have a bad this... but then they are not being used directly AFAIK. Hmm, no they don't or typescript would be complaining that Promise doesn't have a register function - all is ok here.